UCI-IN4MATX-191-Token-ATM/token-atm-spa
# Implement Token ATM Configuration Architecture

### Testing Note

This branch has breaking changes on the Token ATM configuration data structure. Previous Token ATM configuration won't be recognized as valid configuration anymore.

You could use the buttons in Dev-Test page (only shows when Angular is in development mode) to test the architecture. Here are the functionalities of these buttons:

1. Regenerate Content: All content related to Token ATM in the current course will be deleted except for the configuration itself. After that, content (student record encryption key, module, assignment group, log assignment, and request submission quizzes) will be generated again. This will cause all students' request history and token balance being dropped since the log assignment is re-created. This feature is designed to ease testing and migration between courses.
2. Delete All: All content related to Token ATM in the current course will be deleted (including the configuration).
3. Create Configuration: A new Token ATM configuration will be created, while the relevant content (student record encryption key, module, assignment group, and log assignment) will also be created. This feature only works for a course that don't have existing Token ATM configuration (could use "Delete All" to achieve that). This feature is designed to allow teacher initialize Token ATM for a course.
4. Add Token Option Group: A token option group will be added to Token ATM configuration. The request submission quiz for it will also be created, published, and added to module and assignment group. A valid Token ATM configuration is required.
5. Add Token Optionto the Last Token Option Group: A basic token option will be added to the last token option group of the Token ATM Configuration. The request submission quiz will be updated if no one has taken the quiz (see note below). A valid Token ATM configuration and at least one token option group is required.
6. Delete First Token Option Group: The first token option group will be deleted from the Token ATM configuration. The request submission quiz for it will also be deleted. A valid Token ATM configuration and at least one token option group is required.
7. Delete First Token Option in the First Token Option Group: The first token option of the first token option group will be deleted from the Token ATM Configuration. The request submission quiz will be updated if no one has taken the quiz (see note below). A valid Token ATM configuration, at laest one token option group, and at least one token option in the first token option group is required.

You could also modify the code in `dev-test.component.ts` to test it furthermore.

Note: [\[Ref 1\]](https://community.canvaslms.com/t5/Canvas-Question-Forum/Saving-Quizzes-w-API/m-p/226406) [\[Ref 2\]](https://community.canvaslms.com/t5/Canvas-Question-Forum/How-to-save-quiz-via-canvas-API/m-p/480447) Canvas prevents the quiz being automatically and programmatically changed after students have taken it. The teacher has to manually publish the change (by clicking a "Save It Now" button in the quiz management interface). No documented API could get this done automatically, while it's quite hard to reverse engineering it in the remaining time (it seems that Canvas issues a special authenticity_token for it). Also, no documented API could know whether the change is saved or not. The current plan is to popup a modal when the teacher did such a modification to inform them that a manual operation is needed.
